### PR TITLE
Add jni interface to use a binary hnsw index with faiss

### PR DESCRIPTION
**Pull request by heemin32**
### Description
It is safe to merge to main and 2.x as there is no user impact with this PR
 
### Issues Resolved
https://github.com/opensearch-project/k-NN/issues/1767
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by junqiu-lei — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased 2.x](https://github.com/opensearch-project/k-NN/compare/2.15...2.x)
 ### Features
+* JNI layer interface to support Faiss binary index [#1778](https://github.com/opensearch-project/k-NN/pull/1778)
```
Comment (jmazanec15) [inline CHANGELOG.md:17]: eh this isnt a feature if it is not exposed to user. Just skip-change log for this and group it with binary field support feature.

**Review by jmazanec15 — COMMENTED** [review]
overall, looks pretty good. One minor comment on ChangeLog

**Comment by shatejas** [comment]
Why is there a need to merge in main when the feature is not ready? Can't you rebase the changes form main in the feature branch?

File: CHANGELOG.md (hunk 1 above)
Comment (heemin32) [reply CHANGELOG.md:17]: @navneet1v want to add it in change log

**Comment by heemin32** [comment]
> Why is there a need to merge in main when the feature is not ready/users cannot use it? Can't you rebase the changes from main in the feature branch?

This PR can be seen as an isolated works from any feature; Opening up binary index method of faiss through JNI and JNI layer refactoring.

File: CHANGELOG.md (hunk 1 above)
Comment (navneet1v) [reply CHANGELOG.md:17]: This is a feature, but the description of the feature is not correct.

Comment (navneet1v) [reply CHANGELOG.md:17]: Can we make it,
`Added the support for Binary Index in kNNEngine.`

Comment (heemin32) [reply CHANGELOG.md]: Updated. However, I agree with @jmazanec15 that it is not a feature that use can use though.

Comment (heemin32) [reply CHANGELOG.md]: Talked with @navneet1v offline. Will remove the change log entry for this PR.

**Review by navneet1v — DISMISSED** [review]

**Review by jmazanec15 — APPROVED** [review]

**Comment by heemin32** [comment]
Rebased from main branch

**Review by jmazanec15 — APPROVED** [review]

**Review by navneet1v — APPROVED** [review]

**Pull request merged**